TERRAFORM.md

```diff
@@ -0,0 +1,96 @@
+# Terraform Best Practices and Guidelines
+
+General terraform guidelines and best practices
+
+## Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Introduction](#introduction)
+- [Code Organization](#code-organization)
+  - [Reusable Module Usage](#reusable-module-usage)
+- [State Management](#state-management)
+  - [Remote State](#remote-state)
+    - [State key](#state-key)
+  - [State Locking](#state-locking)
+  - [State migration](#state-migration)
+  - [Module Guidelines](#module-guidelines)
+- [Code formatting](#code-formatting)
+
+## Introduction
+
+Terraform is a powerful tool for provisioning and managing infrastructure as code. To ensure maintainability, scalability, and security, it's important to follow best practices and guidelines when working with Terraform.
+
+## Code Organization
+
+The high level code organization should look like this
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- STAGING        # resource pro staging
+```
+
+Then it should mirror the hierarchical structure of the infrastructure. Note that each terraform module (project) should be kept small.
+
+### Reusable Module Usage
+
+Use modules to encapsulate and reuse configurations. This promotes DRY (Don't Repeat Yourself) principles and makes your configurations more manageable.
+
+- Create modules for common resources.
+- Use versioned modules from the Terraform Registry when possible.
+- Use modules to share global variales between terraform modules
+
+## State Management
+
+### Remote State
+
+We use remote state management using AWS S3 backend
+
+- Encrypt the state file at rest (see [Terraform Documentation](https://developer.hashicorp.com/terraform/language/settings/backends/s3#encrypt)).
+- Put the state backend definition in `terraform.tf` file
```

Comment:
Can we enforce that with the CI?

Reply:
With some bash magic we could but this would be complicated in my opinion and I would not try to do it as it might add more trouble than it problem solve.

Reply:
I think it might be useful to enforce this and other conventions in this way but the tools for parsing terraform config seem rather limited at the moment. The supported way is apparently terraform-config-inspect which doesn't even bother surfacing the backend definition. I agree hacking some bash script is likely going to be too fragile to be worth it.